device_e2e/aio/test_send_message_stress.py

```diff
@@ -0,0 +1,597 @@
+        for `test_length_in_seconds` seconds.
+        """
+        test_end = time.time() + test_length_in_seconds
+        loop_index = 0
+        fault_types = list(fault_injection_types.keys())
+
+        while time.time() < test_end:
+            await asyncio.sleep(min(fault_interval, test_end - time.time()))
+
+            if time.time() >= test_end:
+                return
+
+            fault_type = fault_types[loop_index % len(fault_types)]
+            loop_index += 1
+
+            logger.warning("Injecting fault: {}".format(fault_type))
+
+            fault_message = get_fault_injection_message(fault_type)
+
+            await client.send_message(fault_message)
+
+    async def record_stress_measurements(self, client, stop_event, stress_measurements):
+        """
+        Coroutine to record certain test measurement throughout the legnth of
+        a test.  It completes after stop_event is set.  This way, the invoker can control how
+        long this code runs.
+
+        This test records:
+            peak_reconnect_time
+            peak_resident_memory_mb
+        """
+
+        process = psutil.Process(os.getpid())
+        was_connected = True
+        disconnect_time = 0
+
+        def handle_on_connection_state_change():
+            nonlocal was_connected, disconnect_time
+
+            if client.connected:
+                if not was_connected:
+                    reconnect_time = time.time() - disconnect_time
+                    with stress_measurements.lock:
+                        stress_measurements.peak_reconnect_time = max(
+                            stress_measurements.peak_reconnect_time, reconnect_time
+                        )
+                        logger.info(
+                            "Reconnect time = {}, new peak = {}".format(
+                                reconnect_time, stress_measurements.peak_reconnect_time
+                            )
+                        )
+            else:  # not client.connected and
+                if was_connected:
+                    disconnect_time = time.time()
+            was_connected = client.connected
+
+        assert not client.on_connection_state_change
+        client.on_connection_state_change = handle_on_connection_state_change
+
+        while not stop_event.is_set():
+            current_resident_memory_mb = process.memory_info().rss / 1024 / 1024
+            with stress_measurements.lock:
+                stress_measurements.peak_resident_memory_mb = max(
+                    stress_measurements.peak_resident_memory_mb, current_resident_memory_mb
+                )
+                logger.info(
+                    "memory use = {}, new peak = {}".format(
+                        current_resident_memory_mb,
+                        stress_measurements.peak_resident_memory_mb,
+                    )
+                )
+
+            await asyncio.sleep(5)
```

Comment:
May want to note somewhere that the memory use measurements are based on a sample taken every 5 seconds